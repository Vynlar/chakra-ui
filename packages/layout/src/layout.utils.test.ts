import {
  convertToArray,
  fillArray,
  calculateBaselineCrop,
} from "./layout.utils"
import baseTheme from "@chakra-ui/preset-base"

describe("fillArray()", () => {
  const cases = [
    [[], []],
    [["a"], ["a"]],
    [
      ["a", null],
      ["a", "a"],
    ],
    [
      ["a", null, "b"],
      ["a", "a", "b"],
    ],
    [
      [null, "a", null, "b"],
      [null, "a", "a", "b"],
    ],
    [
      ["a", null, "b", null],
      ["a", "a", "b", "b"],
    ],
  ]

  cases.forEach(([input, output], index) => {
    it(`should work for case ${index}`, () => {
      const result = fillArray(input, input.length)
      expect(result).toEqual(output)
    })
  })

  it("should work for longer length", () => {
    const result = fillArray(["a", null, "b"], 5)
    expect(result).toEqual(["a", "a", "b", "b", "b"])
  })

  it("should work for empty array", () => {
    const result = fillArray([], 3)
    expect(result).toEqual([null, null, null])
  })
})

describe("convertToArray()", () => {
  const theme: any = {
    breakpoints: ["A", "B", "C"],
  }
  theme.breakpoints.sm = theme.breakpoints[0]
  theme.breakpoints.md = theme.breakpoints[1]
  theme.breakpoints.lg = theme.breakpoints[2]

  it("should work for a single value", () => {
    expect(convertToArray(theme, 1.3)).toEqual([1.3])
    expect(convertToArray(theme, "34rem")).toEqual(["34rem"])
  })
  it("should work for an array", () => {
    expect(convertToArray(theme, [1.2, "3rem"])).toEqual([1.2, "3rem"])
    expect(convertToArray(theme, [null, "3rem"])).toEqual([null, "3rem"])
  })
  it("should work for an object", () => {
    expect(convertToArray(theme, { sm: 1 })).toEqual([1])
    expect(convertToArray(theme, { lg: 1 })).toEqual([null, null, 1])
    expect(convertToArray(theme, { sm: 1, md: 2 })).toEqual([1, 2])
    expect(convertToArray(theme, { sm: 1, lg: 2 })).toEqual([1, null, 2])
    expect(convertToArray(theme, { md: 1, lg: 2 })).toEqual([null, 1, 2])
  })
})

describe("calculateTextCrop()", () => {
  const theme = {
    ...baseTheme,
    baselineCrop: {
      body: {
        topCrop: 11,
        bottomCrop: 9,
      },
      mono: {
        topCrop: 8.5,
        bottomCrop: 10,
      },
    },
  }

  const cases: Parameters<typeof calculateBaselineCrop>[1][] = [
    {},
    { fontFamily: "mono" },
    { fontFamily: ["body", "mono"] },
    { lineHeight: "normal" },
    { lineHeight: 1.4 },
    { lineHeight: "tall" },
    { lineHeight: ["normal", "tall"] },
    { lineHeight: [null, "tall"] },
    { lineHeight: ["tall", null, "taller", null] },
    { lineHeight: ["tall", 1.5, "taller", null] },
    { lineHeight: ["short", "taller"], fontFamily: ["mono", "body"] },
    {
      lineHeight: ["short", null, "taller"],
      fontFamily: ["mono", "body", null],
    },
  ]

  cases.forEach((props, index) => {
    it(`should work for case #${index + 1}`, () => {
      const result = calculateBaselineCrop(theme, props)
      expect(result).toMatchSnapshot()
    })
  })

  it("array and object styles should produce the same result", () => {
    expect(calculateBaselineCrop(theme, { lineHeight: [1, 2] })).toEqual(
      calculateBaselineCrop(theme, { lineHeight: { sm: 1, md: 2 } }),
    )

    expect(
      calculateBaselineCrop(theme, {
        lineHeight: [1, 2],
        fontFamily: ["mono"],
      }),
    ).toEqual(
      calculateBaselineCrop(theme, {
        lineHeight: { sm: 1, md: 2 },
        fontFamily: { sm: "mono" },
      }),
    )
  })
})
