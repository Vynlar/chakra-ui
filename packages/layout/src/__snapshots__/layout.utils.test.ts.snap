// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculateTextCrop() should work for case #1 1`] = `
Object {
  "textNodeStyles": Object {
    "&::before": Object {
      "content": "\\"\\"",
      "display": "block",
      "marginBottom": Array [
        "-0.48780em",
      ],
    },
    "lineHeight": Array [
      1.2,
    ],
    "transform": Array [
      "translateY(0.21951em)",
    ],
  },
}
`;

exports[`calculateTextCrop() should work for case #2 1`] = `
Object {
  "textNodeStyles": Object {
    "&::before": Object {
      "content": "\\"\\"",
      "display": "block",
      "marginBottom": Array [
        "-0.45122em",
      ],
    },
    "lineHeight": Array [
      1.2,
    ],
    "transform": Array [
      "translateY(0.24390em)",
    ],
  },
}
`;

exports[`calculateTextCrop() should work for case #3 1`] = `
Object {
  "textNodeStyles": Object {
    "&::before": Object {
      "content": "\\"\\"",
      "display": "block",
      "marginBottom": Array [
        "-0.48780em",
        "-0.45122em",
      ],
    },
    "lineHeight": Array [
      1.2,
      1.2,
    ],
    "transform": Array [
      "translateY(0.21951em)",
      "translateY(0.24390em)",
    ],
  },
}
`;

exports[`calculateTextCrop() should work for case #4 1`] = `
Object {
  "textNodeStyles": Object {
    "&::before": Object {
      "content": "\\"\\"",
      "display": "block",
      "marginBottom": Array [
        "-0.48780em",
      ],
    },
    "lineHeight": Array [
      1.2,
    ],
    "transform": Array [
      "translateY(0.21951em)",
    ],
  },
}
`;

exports[`calculateTextCrop() should work for case #5 1`] = `
Object {
  "textNodeStyles": Object {
    "&::before": Object {
      "content": "\\"\\"",
      "display": "block",
      "marginBottom": Array [
        "-0.68780em",
      ],
    },
    "lineHeight": Array [
      1.4,
    ],
    "transform": Array [
      "translateY(0.31951em)",
    ],
  },
}
`;

exports[`calculateTextCrop() should work for case #6 1`] = `
Object {
  "textNodeStyles": Object {
    "&::before": Object {
      "content": "\\"\\"",
      "display": "block",
      "marginBottom": Array [
        "-0.91280em",
      ],
    },
    "lineHeight": Array [
      1.625,
    ],
    "transform": Array [
      "translateY(0.43201em)",
    ],
  },
}
`;

exports[`calculateTextCrop() should work for case #7 1`] = `
Object {
  "textNodeStyles": Object {
    "&::before": Object {
      "content": "\\"\\"",
      "display": "block",
      "marginBottom": Array [
        "-0.48780em",
        "-0.91280em",
      ],
    },
    "lineHeight": Array [
      1.2,
      1.625,
    ],
    "transform": Array [
      "translateY(0.21951em)",
      "translateY(0.43201em)",
    ],
  },
}
`;

exports[`calculateTextCrop() should work for case #8 1`] = `
Object {
  "textNodeStyles": Object {
    "&::before": Object {
      "content": "\\"\\"",
      "display": "block",
      "marginBottom": Array [
        "-0.48780em",
        "-0.91280em",
      ],
    },
    "lineHeight": Array [
      1.2,
      1.625,
    ],
    "transform": Array [
      "translateY(0.21951em)",
      "translateY(0.43201em)",
    ],
  },
}
`;

exports[`calculateTextCrop() should work for case #9 1`] = `
Object {
  "textNodeStyles": Object {
    "&::before": Object {
      "content": "\\"\\"",
      "display": "block",
      "marginBottom": Array [
        "-0.91280em",
        "-0.91280em",
        "-1.28780em",
        "-1.28780em",
      ],
    },
    "lineHeight": Array [
      1.625,
      1.625,
      2,
      2,
    ],
    "transform": Array [
      "translateY(0.43201em)",
      "translateY(0.43201em)",
      "translateY(0.61951em)",
      "translateY(0.61951em)",
    ],
  },
}
`;

exports[`calculateTextCrop() should work for case #10 1`] = `
Object {
  "textNodeStyles": Object {
    "&::before": Object {
      "content": "\\"\\"",
      "display": "block",
      "marginBottom": Array [
        "-0.91280em",
        "-0.78780em",
        "-1.28780em",
        "-1.28780em",
      ],
    },
    "lineHeight": Array [
      1.625,
      1.5,
      2,
      2,
    ],
    "transform": Array [
      "translateY(0.43201em)",
      "translateY(0.36951em)",
      "translateY(0.61951em)",
      "translateY(0.61951em)",
    ],
  },
}
`;

exports[`calculateTextCrop() should work for case #11 1`] = `
Object {
  "textNodeStyles": Object {
    "&::before": Object {
      "content": "\\"\\"",
      "display": "block",
      "marginBottom": Array [
        "-0.62622em",
        "-1.28780em",
      ],
    },
    "lineHeight": Array [
      1.375,
      2,
    ],
    "transform": Array [
      "translateY(0.33140em)",
      "translateY(0.61951em)",
    ],
  },
}
`;

exports[`calculateTextCrop() should work for case #12 1`] = `
Object {
  "textNodeStyles": Object {
    "&::before": Object {
      "content": "\\"\\"",
      "display": "block",
      "marginBottom": Array [
        "-0.62622em",
        "-0.66280em",
        "-1.28780em",
      ],
    },
    "lineHeight": Array [
      1.375,
      1.375,
      2,
    ],
    "transform": Array [
      "translateY(0.33140em)",
      "translateY(0.30701em)",
      "translateY(0.61951em)",
    ],
  },
}
`;
